Technique 'ambient' node case added

poof86 opened the pull request:
The ambient node case is missing from technique/shader parsing

poof86 committed:
Technique 'ambient' node case added

The ambient node case is missing from technique/shader parsing

mrdoob commented:
We don't support `ambient` in our built-in materials though.

poof86 commented:
But there is support for ambient occlusion textures. I was trying to pick up ambient and emissive textures in my models and set them to the material.aoMap property.

Btw, I am missing a lightMap feature for the Basic/Constant material. I see it was removed with version r71.
To still get the result I want (a detailed diffuse texture blended with a lightMap) I changed the THREE.ShaderChunk[ 'aomap_fragment’ ] chunk with the following:

#ifdef USE_AOMAP
	vec3 ambientOcclusion = texture2D( aoMap, vUv2 ).rgb * aoMapIntensity;
	reflectedLight.indirectDiffuse = min( reflectedLight.indirectDiffuse.rgb * 2.0 * ambientOcclusion.rgb, vec3(1.0) );
	#if defined( USE_ENVMAP ) && defined( STANDARD )
		float dotNV = saturate( dot( geometry.normal, geometry.viewDir ) );
		reflectedLight.indirectSpecular *= computeSpecularOcclusion( dotNV, ambientOcclusion, material.specularRoughness );
	#endif
#endif

Inspired by WestLangley comment:
https://github.com/mrdoob/three.js/pull/5741#issuecomment-66477956 <https://github.com/mrdoob/three.js/pull/5741#issuecomment-66477956>

Maybe ColladaLoader.js can use the ambient texture for material.aoMap and emissive texture for material.lightMap?
Lightmaps on BasicMaterial would use the following shader chunk:

#ifdef USE_LIGHTMAP
	reflectedLight.indirectDiffuse = min( reflectedLight.indirectDiffuse.rgb * 2.0 * texture2D( lightMap, vUv2 ).rgb * lightMapIntensity, vec3(1.0) );
#endif


> On 12 Apr 2016, at 13:13, Mr.doob <notifications@github.com> wrote:
> 
> We don't support ambient in our built-in materials though.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/mrdoob/three.js/pull/8610#issuecomment-208850347>

WestLangley commented:
@poof86

Lightmaps in three.js are now treated as a source of indirect (ambient) light, and are purely additive to total indirect light.

Ambient occlusion maps in three.js occlude only indirect (ambient) light.

`MeshBasicMaterial` does not respond to lights.

three.js materials no longer have an `ambient` property. The ambient reflectance of the material is assumed to be the same as the diffuse reflectance of the material (a.k.a, the material's `color`).

poof86 committed:
Allow ColladaLoader.js to set .aoMap property

Have the ColladaLoader.js use emissive and ambient textures to set the aoMap property on BasicMaterials.
In the future maybe add a lightMap property to BasicMaterial and have different blending modes for the shading.

mrdoob commented on examples/js/loaders/ColladaLoader.js:
```diff
@@ -3791,7 +3792,21 @@ THREE.ColladaLoader = function () {
 			case 'constant':
-
+			
+		                if ( props.emissionMap !== undefined ) {
```
Do you have any example files that use these maps?

poof86 replied:
I'll make a quick example just a sec

poof86 commented:
I understand,

It would be nice if there is an easy way to use the BasicMaterial to blend detailed diffuse textures and prebaked lighting textures. Maybe add a different kind of textureMap property for this technique? The BasicMaterial is ideal for when you want to be as less resource intensive as possible right? So support for this would be great :)

Should I close the pull request?

WestLangley commented:
By the way, `MeshBasicMaterial` does support ambient occlusion maps.

https://github.com/mrdoob/three.js/pull/6588
https://github.com/mrdoob/three.js/pull/6364#commitcomment-11288785

mrdoob commented:
Now that `MeshBasicMaterial` supports lightMap. Is this PR needed?

poof86 commented:
The lightmap shader uses a different formula, but I will test to see if it has similar results.
Sorry for not posting an example yet, was quite busy and forgot :/

Mugen87 commented:
Old `ColladaLoader` has been removed.
